# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: netwerk/test/unit/test_standardurl.js
# Commit: 9ad299b5161b
# Full Hash: 9ad299b5161b0cedf852d56365310de499f94043
# Author: Valentin Gosu <valentin.gosu@gmail.com>
# Date: 2018-11-14 22:12:17
# Regressor Bug: 1495313
# File Overlap Count: 1
# Description:
#   Bug 1495313 - Allow empty username in URLs r=dragana
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D11254
# ==============================================================================

diff -r 6293789f6e9c -r 9ad299b5161b netwerk/test/unit/test_standardurl.js
--- a/netwerk/test/unit/test_standardurl.js	Wed Nov 14 14:27:58 2018 +0000
+++ b/netwerk/test/unit/test_standardurl.js	Wed Nov 14 15:24:21 2018 +0000
@@ -544,14 +544,90 @@
   url = url.mutate().setPassword("ppppppppppp").finalize();
   Assert.equal(url.spec, "http://z:ppppppppppp@example.com/");
 
-  url = url.mutate().setUsername("").finalize(); // Should clear password too
-  Assert.equal(url.spec, "http://example.com/");
+  url = stringToURL("http://example.com");
   url = url.mutate().setPassword("").finalize(); // Still empty. Should work.
   Assert.equal(url.spec, "http://example.com/");
 
   run_next_test();
 });
 
+add_test(function test_emptyUser() {
+  let url = stringToURL("http://:a@example.com/path/to/something?query#hash");
+  Assert.equal(url.spec, "http://:a@example.com/path/to/something?query#hash");
+  url = stringToURL("http://:@example.com/path/to/something?query#hash");
+  Assert.equal(url.spec, "http://example.com/path/to/something?query#hash");
+
+  const kurl = stringToURL("http://user:pass@example.com:8888/path/to/something?query#hash");
+  url = kurl.mutate().setUsername("").finalize();
+  Assert.equal(url.spec, "http://:pass@example.com:8888/path/to/something?query#hash");
+  Assert.equal(url.host, "example.com");
+  Assert.equal(url.hostPort, "example.com:8888");
+  Assert.equal(url.filePath, "/path/to/something");
+  Assert.equal(url.query, "query");
+  Assert.equal(url.ref, "hash");
+  url = kurl.mutate().setUserPass(":pass1").finalize();
+  Assert.equal(url.spec, "http://:pass1@example.com:8888/path/to/something?query#hash");
+  Assert.equal(url.host, "example.com");
+  Assert.equal(url.hostPort, "example.com:8888");
+  Assert.equal(url.filePath, "/path/to/something");
+  Assert.equal(url.query, "query");
+  Assert.equal(url.ref, "hash");
+  url = url.mutate().setUsername("user2").finalize();
+  Assert.equal(url.spec, "http://user2:pass1@example.com:8888/path/to/something?query#hash");
+  Assert.equal(url.host, "example.com");
+  url = url.mutate().setUserPass(":pass234").finalize();
+  Assert.equal(url.spec, "http://:pass234@example.com:8888/path/to/something?query#hash");
+  Assert.equal(url.host, "example.com");
+  url = url.mutate().setUserPass("").finalize();
+  Assert.equal(url.spec, "http://example.com:8888/path/to/something?query#hash");
+  Assert.equal(url.host, "example.com");
+  url = url.mutate().setPassword("pa").finalize();
+  Assert.equal(url.spec, "http://:pa@example.com:8888/path/to/something?query#hash");
+  Assert.equal(url.host, "example.com");
+  url = url.mutate().setUserPass("user:pass").finalize();
+  symmetricEquality(true, url.QueryInterface(Ci.nsIURL), kurl);
+
+  url = stringToURL("http://example.com:8888/path/to/something?query#hash");
+  url = url.mutate().setPassword("pass").finalize();
+  Assert.equal(url.spec, "http://:pass@example.com:8888/path/to/something?query#hash");
+  url = url.mutate().setUsername("").finalize();
+  Assert.equal(url.spec, "http://:pass@example.com:8888/path/to/something?query#hash");
+
+  url = stringToURL("http://example.com:8888");
+  url = url.mutate().setUsername("user").finalize();
+  url = url.mutate().setUsername("").finalize();
+  Assert.equal(url.spec, "http://example.com:8888/");
+
+  url = stringToURL("http://:pass@example.com");
+  Assert.equal(url.spec, "http://:pass@example.com/");
+  url = url.mutate().setPassword("").finalize();
+  Assert.equal(url.spec, "http://example.com/");
+  url = url.mutate().setUserPass("user:pass").finalize();
+  Assert.equal(url.spec, "http://user:pass@example.com/");
+  Assert.equal(url.host, "example.com");
+  url = url.mutate().setUserPass("u:p").finalize();
+  Assert.equal(url.spec, "http://u:p@example.com/");
+  Assert.equal(url.host, "example.com");
+  url = url.mutate().setUserPass("u1:p23").finalize();
+  Assert.equal(url.spec, "http://u1:p23@example.com/");
+  Assert.equal(url.host, "example.com");
+  url = url.mutate().setUsername("u").finalize();
+  Assert.equal(url.spec, "http://u:p23@example.com/");
+  Assert.equal(url.host, "example.com");
+  url = url.mutate().setPassword("p").finalize();
+  Assert.equal(url.spec, "http://u:p@example.com/");
+  Assert.equal(url.host, "example.com");
+
+  url = url.mutate().setUserPass("u2:p2").finalize();
+  Assert.equal(url.spec, "http://u2:p2@example.com/");
+  Assert.equal(url.host, "example.com");
+  url = url.mutate().setUserPass("u23:p23").finalize();
+  Assert.equal(url.spec, "http://u23:p23@example.com/");
+  Assert.equal(url.host, "example.com");
+
+  run_next_test();
+});
+
 registerCleanupFunction(function () {
   gPrefs.clearUserPref("network.standard-url.punycode-host");
 });